### ISSUE TITLE
Unhandled Exception

### ISSUE THREAD

**Issue by msgeburt** [akavache/Akavache#343]
Hey.. I can't get Akavache to work.

```
protected override void OnInitialized(){
    InitializeComponent();
    BlobCache.ApplicationName = "MyApp";
    startUp();
}

private async void startUp(){
    string startPage;
    try {
        startPage = await BlobCache.UserAccount.GetObjectAsync("StartPage");
    } catch (KeyNotFoundException ex) {
        startPage = "StartPage";
    }
    await NavigationService.NavigateAsync(startPage);
}
```

I just want to get (if exists) the value of StartPage and then navigate the to stored page.

But every time i run the app i get an unhandled exception with no more details. Am i doing anything wrong?

**Comment by rupreetg** [comment]
This is because you are calling startup method which is having async. The main thread will call the startup method on a different thread, come back in the OnInitialized method and continue executing. While doing this, it doesn't find any NavigationService instance/call (It doesn't know where to go from here now?), it exits and crashes the application. 

Ideal way to have this is - you navigate to the first page and there in the constructor add to cache.

**Comment by ghuntley** [comment]
What is the exception? Where's the stacktrace?

PS: `async void` anywhere in your code base should be considered harmful. Don't do it. 

This knowledge is well documented in blogs, pluralsight, talks and at MSDN!

http://haacked.com/archive/2014/11/11/async-void-methods/

There's one one circumstance where it's permitted. Search your codebase today and make amends :)

**Comment by duhowise** [comment]
Application: UniVoting.Client.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IO.FileLoadException
   at UniVoting.Client.App+<OnStartup>d__1.MoveNext()
   at System.Runtime.CompilerServices.AsyncVoidMethodBuilder.Start[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.__Canon ByRef)
   at UniVoting.Client.App.OnStartup(System.Windows.StartupEventArgs)
   at System.Windows.Application.<.ctor>b__1_0(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
   at System.Windows.Threading.Dispatcher.PushFrame(System.Windows.Threading.DispatcherFrame)
   at System.Windows.Application.RunDispatcher(System.Object)
   at System.Windows.Application.RunInternal(System.Windows.Window)
   at System.Windows.Application.Run(System.Windows.Window)
   at System.Windows.Application.Run()
   at UniVoting.Client.App.Main()


same error. and i don't event have an async void in my code except for the event handlers

**Comment by duhowise** [comment]
removed Async using blocking calls on all functions at startup 

_positions = new List<Position>();

			try
			{
				var electionData = ElectionConfigurationService.ConfigureElection(Convert.ToInt32(Settings.Default.ElectionId));
				BlobCache.UserAccount.InsertObject("ElectionSettings", electionData).Wait();
				_positions = ElectionConfigurationService.GetAllPositions();

				BlobCache.UserAccount.InsertObject("ElectionPositions", _positions).Wait();
			}
			catch (Exception exception)
			{
				MessageBox.Show(exception.Message, " Election Settings Error");
			}


still crashing with the above error

**Comment by duhowise** [comment]
`_positions = new List<Position>();
try
{
var electionData=ElectionConfigurationService.ConfigureElection(Convert.ToInt32(Settings.Default.ElectionId));
BlobCache.UserAccount.InsertObject("ElectionSettings", electionData).Wait();
_positions = ElectionConfigurationService.GetAllPositions();

BlobCache.UserAccount.InsertObject("ElectionPositions", _positions).Wait();
}
catch (Exception exception)
{
	MessageBox.Show(exception.Message, " Election Settings Error");
}`

**Comment by Im-PJ** [comment]
```
  BlobCache.Secure.GetObject<User>(AppConstants.UserKey)
                               .Subscribe(x =>
                               {
                                   if (x != null && !string.IsNullOrEmpty(x.session_id))
                                       NavigationService.NavigateAsync("/MainPage/Navigation/UserProfile");                              
                               }, 
                               ex => {
                                   NavigationService.NavigateAsync("LoginPage");
                               });
```
I'm also seeing this crash. What is the right way to check if the user is logged in via akavache and navigate to page accordingly?

**Comment by marcohajek** [comment]
@Im-PJ are you sure that the navigation is correct?  Is the view registered for navigation?

Otherwise you can try it with a simpler way:

var x = await BlobCache....

if x != null && ...

just do it like that is a bit easier i guess..

Sorry for the bad formstting im writing thos comment on my phone

**Comment by PureWeen** [comment]
@Im-PJ   it looks like you're using Xamarin Forms and Prism? Is that correct?

Some thoughts
1) You're just calling NavigateAsync but not awaiting the task
2) Xamarin forms is really pesky when it comes to startup :-(  From my experience you have to do everything synchronously up until you provide a page to Xamarin Forms otherwise it can throw odd exceptions.  I've found it's best to just have a landing page you start with and then work from there. I know that's not ideal but until Xamarin Forms provides some sort of  async way to load that first page there's not great options.  
3) I'm not sure how prism handles this but when you get data back from Akavache you will most likely be on a TaskPool  thread (inside the subscribe or error handler) and not a UI Thread. Can Prism handle Navigation that's instigated off the UI Thread?

If you look here you'll see one way I've tried is to just Wait on the observable
https://github.com/PureWeen/ReactiveUI.XamlForms.Sample/blob/master/ReactiveUI.XamlForms.Sample/App.xaml.cs

I don't really do it that way these days and will probably swap that sample out but it's an example.

Something to keep in mind also is that there's a load penalty to Akavache as well so I've found you can shave off between 1 to 2 seconds not touching Akavache until after that first page loads

**Comment by Im-PJ** [comment]
@PureWeen  Yes I'm using XF and Prism. You were spot on with 3rd pt. 
This worked   `Device.BeginInvokeOnMainThread(() => NavigationService.NavigateAsync("LoginPage"));`
I will go through your example to learn more about Rx. 

> Something to keep in mind also is that there's a load penalty to Akavache as well so I've found you can shave off between 1 to 2 seconds not touching Akavache until after that first page loads
I was not aware of this. Still looking for best way to check whether user is logged in, during application startup.

**Comment by PureWeen** [comment]
Cool glad that worked

Yea I gave up trying to make it work during application startup :-(  The main barrier I was hitting is that there is a bug in iOS 8 with deep linking that adds about 5 seconds on your startup because of iOS things

https://techcrunch.com/2016/03/28/its-not-just-you-clicking-on-links-in-ios-9-3-can-crash-your-iphone/

So on my test device (iphone4 and ios 8.4) I was just trimming things down as much as I could in order to get deep linking not crashing on an ios 8 device and just getting that first page to load ASAP before loading anything else.

For example I'm using Firebase.  I found that the configure calls added around a second so even for Firebase I just pushed the configure calls to after the first page loads.  Then on my first page load I just have a "we are loading your data" message at which point I will swap out the root page if they are logged in.   I know it sounds lame in cases where they are logged in but I've found that it feels faster for the user to just load a loading page ASAP then load the right page.

**Comment by Im-PJ** [comment]
Sounds interesting. Just to clarify, I initiliase everything in App.cs in
xamarin forms.
So do you mean to move all initialise code to Page1 and keep app.cs lean?
I also have splashactivity for android.

It would have been easier and encouraging to see some code.

**Comment by PureWeen** [comment]
Yea that's what I'm currently doing and I have no credible reference to say it's the best way to do it but it's what I've found works for me to get to that first page load quickly. I have a Splash Screen on iOS as well but iOS doesn't care :-) you hit that load time limit and it will just kill the app

My plan is to eventually convert my sample over to something to be proud of but in my current app basically i just do it like this (I'm using ReactiveUI)

I've extracted out what I'm doing and simplified it down a bit here so hopefully this makes sense


```
public class MyApplicationRootNavigationPage
...
            this.WhenActivated((CompositeDisposable disp) =>
            {
                if (!ApplicationStateService.Current.AppLoaded)
                {
                   BlobCache.ApplicationName = "AppName";
                    ApplicationStateService.Current.AppLoaded = true;

                   //I wait to even instantiate any VMs until the first page is loaded on screen
                    MainPageViewModel viewModel = new MainPageViewModel();
                    BindingContext = viewModel;
                }
            }
```

Then I just expose that as a BehaviorSubject from ApplicationStateService.Current

```
 bool _AppLoaded = false;
        BehaviorSubject<bool> _appLoadedSignal = new BehaviorSubject<bool>(false);

        public bool AppLoaded
        {
            get { return _AppLoaded; }
            set { _AppLoaded = value; _appLoadedSignal.OnNext(value);}
        }


        public IObservable<bool> AppLoadedSignal { get { return _appLoadedSignal.AsObservable().DistinctUntilChanged().Where(x=> x).Take(1); } }
```

Then I just wire stuff to it so in my finishedLaunching on iOS I have

```
 ApplicationStateService
	                .Current
	                .AppLoadedSignal
	                .ObserveOn(RxApp.TaskpoolScheduler)
	                .Subscribe(
	                _ =>
	                {
						Firebase.Analytics.Options.DefaultInstance.DeepLinkUrlScheme = "SCHEME";
						Firebase.Analytics.App.Configure();		                    
	                });
```


Basically anything that isn't needed to setup that first page I just delay until that FirstPage is visible

**Comment by Inrego** [comment]
This is still quite a relevant issue, as there's no async way to initialize app. I too want to check if user is logged (by reading data from akavache), before I decide whether to show login screen or home screen.
Is a "second splash screen" still really the way to go?

**Comment by glennawatson** [comment]
Thank you for your report. Closing as this is no longer applicable to Akavache V11.

V11 no longer uses PCL (Portable Class Libraries) or the old static `BlobCache.ApplicationName` initialization. The new initialization uses a builder pattern:

```csharp
CacheDatabase.Initialize<SystemJsonSerializer>(builder =>
    builder.WithApplicationName("MyApp")
           .WithSqliteProvider()
           .WithSqliteDefaults());

var startPage = await CacheDatabase.UserAccount.GetObject<string>("StartPage");
```

V11 targets modern .NET (net8.0+) and .NET MAUI, with explicit serializer and provider configuration that avoids the silent initialization failures that occurred with the old API.

If you encounter issues with V11, please open a new issue with the V11-specific details.
